[22851] Table Inline Create

oliverguenther opened the pull request:
Adds a new row for allowed users to create a work package within the table view.

Steps / TODOS
- [x] POST empty /api/v3/work_packages/form with first type for empty resource
- [x] Add the resulting wp resource to rows
- [x] Activate inline edit for all required fields
- [x] Focus into first required field
- [x] When pressing <enter>, create a new line starting with the above.

_Out of scope_ (handled in separate WP).
- [ ] Use Global post even in project context.
- [ ] Handle empty tables

oliverguenther committed:
Create new button directive and place in table

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-inline-create-row.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-inline-create-row.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

furinvader committed:
Move angular module definitions into separate file

oliverguenther commented:
https://community.openproject.com/work_packages/22851/activity

furinvader committed:
Move the WP create button to the WP buttons directory

furinvader committed:
Move the keepTab service to the wp-panels directory

furinvader committed:
Move the WP create button into a separate file

furinvader committed:
Use TS in WP create button directive

furinvader committed:
Refactor the WP create button controller

furinvader committed:
Move the has-context-menu directive to the components directory

HDinger commented:
I opened a PR which places the create button in the footer of the WP table. 
https://github.com/oliverguenther/openproject/pull/4

furinvader committed:
Refactor has-context-menu item focus

furinvader committed:
Export and use wpButton module

oliverguenther committed:
Add title to form properties

oliverguenther committed:
Add -new modifier to new work package rows

oliverguenther committed:
Reduce error widths to cell span

oliverguenther committed:
Save or update work package conditionally

oliverguenther committed:
Mark work package as new

furinvader committed:
Add WP inline create directive

furinvader committed:
Implement the WP inline create button

oliverguenther committed:
Style button correctly

oliverguenther committed:
Decide when to activate the field for new WPs

oliverguenther committed:
Remove padding on -new row

oliverguenther committed:
Avoid animation on added -new rows

oliverguenther committed:
Avoid writing title when passed from new workPackage

oliverguenther committed:
Remove jqMigrate warning

oliverguenther committed:
Avoid animations for newly added rows

oliverguenther committed:
Add cancel button to inline create row

oliverguenther committed:
Track new work packages by custom id

If we don't track new work packages, ng-animate will freak out due to
the track by expression. We thus track work packages by setting a custom id set from the
current date.

oliverguenther committed:
Disable focus on errors outside accessibility mode

oliverguenther committed:
Add inline create cancel context menu

oliverguenther committed:
Conditionally set focus on fields

Set focus on the first active field

furinvader commented on frontend/app/components/wp-buttons/wp-create-button/wp-create-button.directive.ts:
```diff
@@ -42,31 +40,8 @@ function wpCreateButton() {
-    });
-  }
-}
+wpButtonsModule.directive('wpCreateButton', wpCreateButton);
```
@romanroe Here is an example, where we make use of the exported modules.

furinvader commented:
Closed, as it is incorporated into #4290

oliverguenther committed:
Fix TS and CC warnings

oliverguenther committed:
Fix tests
